src/ssl.c

```diff
@@ -49672,40 +50281,112 @@ char* wolfSSL_BN_bn2dec(const WOLFSSL_BIGNUM* bn)
 }
 #endif /* defined(WOLFSSL_KEY_GEN) || defined(HAVE_COMP_KEY) */
 
-/* return code compliant with OpenSSL :
- *   1 if success, 0 else
+/* Internal function for adding/subtracting an unsigned long from a
+ * WOLFSSL_BIGNUM. To add, pass "sub" as 0. To subtract, pass it as 1.
+ * Returns 1 (WOLFSSL_SUCCESS) on success and 0 (WOLFSSL_FAILURE) on failure.
  */
-int wolfSSL_BN_lshift(WOLFSSL_BIGNUM *r, const WOLFSSL_BIGNUM *bn, int n)
+static int wolfSSL_BN_add_word_int(WOLFSSL_BIGNUM *bn, WOLFSSL_BN_ULONG w,
+                                   int sub)
 {
-    WOLFSSL_MSG("wolfSSL_BN_lshift");
+    int ret = WOLFSSL_SUCCESS;
+    int rc = 0;
+    mp_int w_mp;
 
-    if (r == NULL || r->internal == NULL || bn == NULL || bn->internal == NULL){
+    if (bn == NULL || bn->internal == NULL) {
         WOLFSSL_MSG("bn NULL error");
-        return WOLFSSL_FAILURE;
+        ret = WOLFSSL_FAILURE;
     }
 
-    if (mp_mul_2d((mp_int*)bn->internal, n, (mp_int*)r->internal) != MP_OKAY) {
-        WOLFSSL_MSG("mp_mul_2d error");
-        return WOLFSSL_FAILURE;
+    if (ret == WOLFSSL_SUCCESS) {
+        if (w <= MP_MASK) {
+            if (sub == 1) {
+                rc = mp_sub_d((mp_int*)bn->internal, (mp_digit)w,
+                              (mp_int*)bn->internal);
+            }
+            else {
+                rc = mp_add_d((mp_int*)bn->internal, (mp_digit)w,
+                              (mp_int*)bn->internal);
+            }
+            if (rc != MP_OKAY) {
+                WOLFSSL_MSG("mp_add/sub_d error");
+                ret = WOLFSSL_FAILURE;
+            }
+        }
+        else {
+            if (mp_init(&w_mp) != MP_OKAY) {
+                ret = WOLFSSL_FAILURE;
+            }
+            if (ret == WOLFSSL_SUCCESS) {
+                if (mp_set_int(&w_mp, w) != MP_OKAY) {
+                    ret = WOLFSSL_FAILURE;
+                }
+            }
+            if (ret == WOLFSSL_SUCCESS) {
+                if (sub == 1) {
+                    rc = mp_sub((mp_int *)bn->internal, &w_mp,
+                                (mp_int *)bn->internal);
+                }
+                else {
+                    rc = mp_add((mp_int *)bn->internal, &w_mp,
+                                (mp_int *)bn->internal);
+                }
+                mp_free(&w_mp);                
+                if (rc != MP_OKAY) {
+                    WOLFSSL_MSG("mp_add/sub error");
+                    ret = WOLFSSL_FAILURE;
+                }
+            }
+        }
     }
 
-    return WOLFSSL_SUCCESS;
+    return ret;
 }
 
 /* return code compliant with OpenSSL :
  *   1 if success, 0 else
  */
-int wolfSSL_BN_rshift(WOLFSSL_BIGNUM *r, const WOLFSSL_BIGNUM *bn, int n)
+int wolfSSL_BN_add_word(WOLFSSL_BIGNUM *bn, WOLFSSL_BN_ULONG w)
```

Comment:
Body of this function was moved to `wolfSSL_BN_add_word_int` and adapted to support subtraction, too. Now `wolfSSL_BN_add_word_int` and `wolfSSL_BN_sub_word` just wrap that function.